spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/route/CachingRouteLocator.java

```diff
@@ -89,25 +89,25 @@ public void onApplicationEvent(RefreshRoutesEvent event) {
 						.onErrorResume(s -> Mono.just(List.of()));
 
 				scopedRoutes.subscribe(scopedRoutesList -> {
-					Flux.concat(Flux.fromIterable(scopedRoutesList), getNonScopedRoutes(event))
-							.sort(AnnotationAwareOrderComparator.INSTANCE).materialize().collect(Collectors.toList())
-							.subscribe(this::publishRefreshEvent, this::handleRefreshError);
+					updateCache(Flux.concat(Flux.fromIterable(scopedRoutesList), getNonScopedRoutes(event))
+							.sort(AnnotationAwareOrderComparator.INSTANCE));
 				}, this::handleRefreshError);
 			}
 			else {
 				final Mono<List<Route>> allRoutes = fetch().collect(Collectors.toList());
-
-				allRoutes.subscribe(
-						list -> Flux.fromIterable(list).materialize().collect(Collectors.toList())
-								.subscribe(this::publishRefreshEvent, this::handleRefreshError),
-						this::handleRefreshError);
+				allRoutes.subscribe(list -> updateCache(Flux.fromIterable(list)), this::handleRefreshError);
 			}
 		}
 		catch (Throwable e) {
 			handleRefreshError(e);
 		}
 	}
 
+	private synchronized void updateCache(Flux<Route> routes) {
```

Comment:
Should I add some comments as to why is synchronized or the Git history should sufice?